### ISSUE TITLE
Visual block insert – preview results

### ISSUE THREAD

**Issue by amarakon** [neovim/neovim#20329]
### Feature already in Vim?

No

### Feature description

Since you can preview the results of `:s/` for example, I think you should also be able to preview the results of visual block inserting. Let's say we have this block of text:

```
text1
text2
text3
```

At the top left, I press <kbd>Control-V</kbd> to enter visual block mode. I go down twice and then press <kbd>Shift-I</kbd>. At this point I should be able to see three cursors in front of each line. As I type some text, I should be seeing the text on all of the lines, not just the top one. This should definitely be the default behavior, just like `incsearch`.

**Comment by amarakon** [comment]
Also, the same should be true if you use a command like `4i` which will duplicate your insertion four times.

**Comment by TheLeoP** [comment]
I'm interested in trying to tackle this issue. I've been exploring the code and it seems like, while on insert mode, the buffer is updated in real time. And, when leaving insert mode, the text for the other affected locations is computed and insured.

This differs from `:s`, where the changes are made only when the command is submitted. So, the changes made by the preview function can be discarded.

So, what would be the preferred approach in this case? I can think of:

- keep inserting the visual block text while leaving insert mode. Add a preview while typing. 
    - would this imply that the real time insert mode text would be a preview and would need to be inserted later? 
- update the visual block text in real time
    - the text should be computed early
    - i think this should be handed as a special case in the functions that handle insert mode changes
    - the code for inserting the visual block text while leaving insert mode should be removed
- some other approach (?

Also, is the code diverging from its Vim counterpart a concern?

**Comment by justinmk** [comment]
> is the code diverging from its Vim counterpart a concern?

no

> keep inserting the visual block text while leaving insert mode

I don't understand what this means.

> This differs from `:s`, where the changes are made only when the command is submitted.

`:s` actually performs the edits. Then it (1) uses undo to discard the changes, and (2) prunes the undo-tree.

For visual block changes, I suggest this approach:

1. instead of waiting until the end to apply the changes, just do it immediately, on each keypress
2. if user cancels via `ctrl-c`, perform an undo + prunetree (similar to `:s` preview)

**Comment by TheLeoP** [comment]
How should we handle newlines? Currently, doing `<c-v>jIsomething<esc>` results in

```
something
something
```

but doing `<c-v>jIsomething<cr><esc>` results in

```
something

```

Should the same thing happen when upsetting the text in real time? I guess it should be possible using undo.

**Comment by justinmk** [comment]
I would expect that we just call the current code earlier, so it behaves exactly the same as it already does.

**Comment by TheLeoP** [comment]
I don't think that's 100% possible, I'm gonna give a few examples I've encountered while tinkering with a naive implementation.

Since the current implementation

1. Gets the newly inserted text when exiting insert mode

https://github.com/neovim/neovim/blob/665a0e85c4788cb2847e270c333c0aee306f07ad/src/nvim/ops.c#L2448

2. Inserts it in the respective places

https://github.com/neovim/neovim/blob/665a0e85c4788cb2847e270c333c0aee306f07ad/src/nvim/ops.c#L2473-L2475

I've encountered the following problems:

## backspaces

It doesn't handle `<bs>`, `<c-w>`, `<c-u>` or any other form of insert mode text deletion because it doesn't need to. When inserting text in real time this would be necessary. I'm my on-progress naive implementation I'm handling it  in `del_bytes`

https://github.com/neovim/neovim/blob/665a0e85c4788cb2847e270c333c0aee306f07ad/src/nvim/change.c#L871

## newlines

It doesn't do anything if the cursor moved to a different line

https://github.com/neovim/neovim/blob/665a0e85c4788cb2847e270c333c0aee306f07ad/src/nvim/ops.c#L2372-L2374

It would have to undo only the visual block inserted text to keep the current behavior.

I was trying the idea of simply disabling visual block insert when a newline is entered and continue as regular insert mode, hence my question.

**Comment by TheLeoP** [comment]
Now that I understand your expectations and the code better, I think this could be done more like how `:s` works. Saving the undo information, performing the changes on each keystroke and restoring the undo information.

I guess I could do it inside of `insert_handle_key`

https://github.com/neovim/neovim/blob/665a0e85c4788cb2847e270c333c0aee306f07ad/src/nvim/edit.c#L705

**Comment by justinmk** [comment]
> It doesn't handle `<bs>`, `<c-w>`, `<c-u>` or any other form of insert mode text deletion

Yeah, one approach would be to replay each key similar to a macro. See also [paste_store](https://github.com/neovim/neovim/blob/0bef3b911cc262a007fb4412d864c1832d1268ad/src/nvim/getchar.c#L3223) for a related concept.



> I think this could be done more like how `:s` works

SGTM. With test coverage, we can always change the implementation later.

**Comment by phanen** [comment]
FYI: https://github.com/phanen/vbi.nvim as a workaround before it land in core.
